captum/attr/_core/latent_shift.py

```diff
@@ -0,0 +1,386 @@
+
+        if fix_range:
+            lbound, rbound = fix_range
+        else:
+            # Left range
+            lbound = 0
+            last_pred = initial_pred
+            while True:
+                x_lambdax, cur_pred = compute_shift(lbound)
+                pixel_diff = torch.abs(x_lambda0 - x_lambdax).sum().detach()
+
+                # If we stop decreasing the prediction
+                if last_pred < cur_pred:
+                    break
+                # If the prediction becomes very low
+                if cur_pred < 0.05:
+                    break
+                # If we have decreased the prediction by pred_diff
+                if initial_pred - search_pred_diff > cur_pred:
+                    break
+                # If we are moving in the latent space too much
+                if lbound <= -search_max_steps:
+                    break
+                # If we move too far we will distort the image
+                if pixel_diff > search_max_pixel_diff:
+                    break
+
+                last_pred = cur_pred
+                lbound = lbound - search_step_size + lbound // 10
+
+            # Right range search not implemented
+            rbound = 0
+
+        if verbose:
+            print("Selected bounds: ", lbound, rbound)
+
+        # Sweep over the range of lambda values to create a sequence
+        lambdas = np.arange(
+            lbound, rbound, np.abs((lbound - rbound) / lambda_sweep_steps)
+        ).tolist()
+
+        preds = []
+        generated_images = []
+
+        for lam in lambdas:
+            x_lambdax, pred = compute_shift(lam)
+            generated_images.append(x_lambdax.cpu().numpy())
+            preds.append(float(pred))
+
+        params = {}
+        params["generated_images"] = generated_images
+        params["lambdas"] = lambdas
+        params["preds"] = preds
+
+        x_lambda0 = x_lambda0.detach().cpu().numpy()
+        if heatmap_method == "max":
+            # Max difference from lambda 0 frame
+            heatmap = np.max(np.abs(x_lambda0[0][0] - generated_images[0][0]), 0)
+
+        elif heatmap_method == "mean":
+            # Average difference between 0 and other lambda frames
+            heatmap = np.mean(np.abs(x_lambda0[0][0] - generated_images[0][0]), 0)
+
+        elif heatmap_method == "mm":
+            # Difference between first and last frames
+            heatmap = np.abs(generated_images[0][0][0] - generated_images[-1][0][0])
+
+        elif heatmap_method == "int":
+            # Average per frame differences
+            image_changes = []
+            for i in range(len(generated_images) - 1):
+                image_changes.append(
+                    np.abs(generated_images[i][0][0] - generated_images[i + 1][0][0])
+                )
+            heatmap = np.mean(image_changes, 0)
+        else:
+            raise Exception("Unknown heatmap_method for 2d image")
+
+        params["heatmap"] = heatmap
+
+        return params
```

Comment:
I refactored it so it will return a single batched result of heatmaps. If a return_dicts argument is set to true it will return a list of dicts with the details of the computation (needed for videos). What do you think?